OVDB-136: Add AttributeSet::Info convenience class

danrbailey opened the pull request:
This new class resolves a number of issues:

* It collates information that lives in the attribute set descriptor (the names, types and order of attributes) with information that lives solely on the attribute arrays (flag state, strides, etc). This is a frequent source of complexity as it involves navigating both data structures simultaneously to fully understand the array data.
* Allows for turning the existing AttributeSet constructors into delegating constructors and adds a new constructor that contains all the logic for initializing the new AttributeSet so this lives in one place.
* Provides a way to steal attribute array data from an existing attribute set during construction, as an intentional design decision means that there are no methods for resizing attribute arrays once created. 
* Offers a clean mechanism for merging many AttributeSets into one - you create AttributeSet::Info classes out of each and merge those (as they contain all required information to do so), then create a new AttributesSet from the final AttributeSet::Info. @Idclip note - this method is purely a placeholder right now, it needs #631 to be merged to share this portion.

(@richhones @kubaroth)

danrbailey commented on openvdb/points/AttributeSet.cc:
```diff
@@ -1353,10 +1369,118 @@ AttributeSet::Descriptor::read(std::istream& is)
+    for (const auto& namePos : mDescriptor->groupMap()) {
+        os << "\t\t[" << namePos.second << "] " << namePos.first << std::endl;
+    }
+    os << "\tMetadata: " << mDescriptor->getMetadata().metaCount() << " metadata element(s)" << std::endl;
```
Not sure this is particularly performance-critical, but I fixed this anyway. Thanks.

Idclip replied:
Doubt it, but why would we not!

kubaroth replied:
According to https://en.cppreference.com/w/cpp/io/manip/endl

> In many implementations, standard output is line-buffered, and writing '\n' causes a flush anyway ... unnecessary endl only degrades the performance of file output, not standard output.

kubaroth commented on openvdb/points/AttributeSet.cc:
```diff
@@ -79,45 +75,57 @@ AttributeSet::AttributeSet(const AttributeSet& attrSet, Index arrayLength,
-        const AttributeArray* existingArray = attrSet.getConst(pos);
-        const bool constantStride = existingArray->hasConstantStride();
-        const Index stride = constantStride ? existingArray->stride() : existingArray->dataSize();
+        bool constantStride = arrayInfo.constantStride;
```
Was the `const` dropped on purpose in `bool constantStride`?

danrbailey replied:
Thanks, fixed.

kubaroth commented:
Hey Dan, I'm taking another look at the list of members of the AttributeSet::Array struct and was wondering if adding following would require much work?
- type
- isUniform
- codec
- shared

danrbailey commented:
> Hey Dan, I'm taking another look at the list of members of the AttributeSet::Array struct and was wondering if adding following would require much work?
> 
> * type
> * isUniform
> * codec
> * shared
>   EDIT: Perhaps I'm missing that the above ones can be accessed with a descriptor?

The type and codec values can be retrieved from the descriptor - `const NamePair& Descriptor::type(size_t pos) const`. Think Nick already commented to that effect, but the first value is the value type, the second is the codec.

Whether the array is shared or not, you can retrieve from the attribute set - `AttributeSet::isShared(size_t pos)`. However, for the print binary, I'm not sure it's necessary to include this because it'll never be true - there's no concept of instancing when the grids are deserialized so an attribute array couldn't be shared across multiple leaf nodes or grids.

That leaves whether an array is uniform or not. I hesitated a little on whether to include this, but decided not to. The primary purpose of this new data structure is for attribute set construction and for this purpose it's redundant. Arrays should always be constructed as uniform and then expanded when first populated (either implicitly through attempting to set values or explicitly). Providing a new mechanism for producing a non-uniform array would be unnecessarily confusing I think. I don't think this feature is out of place in the print binary so that would mean having to iterate over all the arrays, apologies I didn't think of this when I first proposed that this class would alleviate this.

kubaroth commented:
Thanks for clarification.

danrbailey commented:
I don't believe there to be any more feedback needing addressed for this PR. @Idclip - if you're happy can you approve please?

danrbailey commented:
> I don't believe there to be any more feedback needing addressed for this PR. @Idclip - if you're happy can you approve please?

@Idclip - can you have a last look over my changes please? I believe I've addressed your feedback now.

Idclip commented on openvdb/points/AttributeSet.cc:
```diff
@@ -1353,10 +1369,119 @@ AttributeSet::Descriptor::read(std::istream& is)
+AttributeSet::Info::Info(const AttributeSet::Descriptor::Ptr& descriptor)
+    : mDescriptor(descriptor)
+    , mArrayInfo(descriptor->size(), Array())
+{
```
What's the point on this constructor? AS far as I can tell it currently serves no purpose and makes the API more confusing when using this class.

danrbailey replied:
There's an `AttributeSet(const Descriptor::Ptr& descriptor)` constructor. Introducing a matching constructor here means that all AttributeSet construction can be done from an Info instance which keeps the logic in one place.

I consider it an advanced option to build and manipulate an Info class yourself for AttributeSet construction. For most use cases, I expect users to continue to use the other AttributeSet constructors directly.

Idclip commented on openvdb/points/AttributeSet.h:
```diff
@@ -540,6 +558,82 @@ class OPENVDB_API AttributeSet::Descriptor
+/// AttributeSets.
+/// @note For performance reasons, this class uses copy-on-write for storing
+/// the Descriptor so any modification of the source Descriptor also requires
+/// a deep-copy for the data in this class to remain valid.
```
This comment doesn't make sense. Are you trying to say that if modify the attributes you need to rebuild an instance of this class?

danrbailey replied:
Yes exactly. I agree the wording can be improved here.

Idclip commented on openvdb/points/AttributeSet.h:
```diff
@@ -540,6 +558,82 @@ class OPENVDB_API AttributeSet::Descriptor
+    /// @brief Return an Array metadata object by name
+    /// @param name The name of the array
+    /// @throws LookupError if name does not exist
+    Array& arrayInfo(const Name& name);
```
What's the purpose of returning modifiable Array objects with any of these member methods?

danrbailey replied:
You can use it in combination with the `AttributeSet(const Info& ...)` constructor to compose exactly how you'd like to construct the desired AttributeSet. I should highlight that there's no link between this `Info` data and the member descriptor. It's purely transitory metadata that is passed down to the constructed AttributeArray.

Idclip commented on openvdb/points/AttributeSet.h:
```diff
@@ -540,6 +558,82 @@ class OPENVDB_API AttributeSet::Descriptor
+    /// @brief Return the descriptor
+    const Descriptor& descriptor() const { return *mDescriptor; }
+    /// @brief Return a shared pointer to the descriptor
+    Descriptor::Ptr descriptorPtr() const { return mDescriptor; }
```
Do we really need all 3 of these? The one which returns a ptr provides all access types

danrbailey replied:
I don't see a problem with doing this. There's a (small) performance penalty with copying the shared ptr, you use the one that's most appropriate.

There's also precedence for exposing multiple different access methods elsewhere in the codebase (`Grid::baseTreePtr()`, `Grid::baseTree()`, `Grid::constBaseTree()` for example).

Idclip commented on openvdb/points/AttributeSet.h:
```diff
@@ -69,7 +72,19 @@ class OPENVDB_API AttributeSet
+    /// cleared on completion of this constructor
+    AttributeSet(const Info& info, Index arrayLength,
+        const AttributeArray::ScopedRegistryLock* lock = nullptr,
+        AttributeSet* attributeSet = nullptr);
```
I don't understand the purpose of the info class here. It seems that the same behaviour could be achieved with a bool `steal` flag. I'm not sure the benefit of consolidating the code justifies how convoluted this makes it (i.e. an info class which is expected to hold a descriptor also matched by a provided attribute set). You could, for example, achieve the same consolidation with a different function.

danrbailey replied:
It cannot be achieved with a new `steal` flag - the current constructor takes a const reference which is what is expected of a copy constructor. Even making the reference non-const would be confusing. There's no simple solution here in my opinion and adding another constructor with largely duplicate logic from the other two is not ideal. I don't even think the status quo is a good solution - having two similar, but subtly different pieces of logic for AttributeSet construction is not ideal in my opinion.

I consider the `Info` instance to be an implementation detail of the `AttributeSet(const AttributeSet&)` and `AttributeSet(const DescriptorPtr& descriptor)`, it's not required to know the internal mechanism for construction when using this API. I consider the `AttributeSet(const Info&,...)` constructor to be the advanced constructor, which you'd only use in certain specific cases.

Idclip commented on openvdb/points/AttributeSet.h:
```diff
@@ -540,6 +558,82 @@ class OPENVDB_API AttributeSet::Descriptor
+
+    /// Placeholder method for merging another Info object into this one
+    /// @throws NotImplementedError
+    void merge(const AttributeSet::Info&);
```
Could just not introduce this now?

danrbailey replied:
If possible, I would prefer to make progress on agreeing on the public API, before we move on to discussing this quite complex method that has a lot of implementation detail in it. (I merely included it here to help illustrate the intention of this class).

Idclip commented on openvdb/points/AttributeSet.h:
```diff
@@ -540,6 +558,82 @@ class OPENVDB_API AttributeSet::Descriptor
+    void print(std::ostream& os = std::cout) const;
+
+private:
+    Descriptor::Ptr mDescriptor = std::make_shared<Descriptor>();
```
At first glance, this class seems to me to be a "viewer" class which helps consolidate data from both the attribute set and descriptor. I question why the descriptor that's stored here is modifiable through the member methods. This introduces another layer of complexity. It seems that methods like `merge` would do better to return new instances of the descriptor.

danrbailey replied:
It does increase the complexity, but it's primarily done like this for performance reasons. Doing what you're suggesting doesn't scale to wanting to merge hundreds or even thousands of Info classes or Descriptors.

This class is serving two different use cases - however I consider the "viewer" use case to be mainly a side benefit, the real purpose is a composable description of what you want the AttributeSet to contain. (I did consider two different classes, one for each use case, but I preferred the simplicity of putting everything in one class). 

I would be open to changing the name if you think we can make this primary use-case clearer, `Info` is a pretty generic name.

richhones commented:
I finally got round to looking at this following it being raised at the last TSC meeting. I think that similar to @kubaroth 's comments asking for the type and codec and @Idclip 's comments about storing the Descriptor::Ptr that I think this doesn't really go all the way of being properly convenient for the general case as this been designed specifically for the case of what you needed for merging. Either it should go further to simplify the AttributeSet/Descriptor stuff - my suggestion would be to remove the Descriptor from the Info class and store the type and codec information in the Array struct. Alternatively just put this internal to the merging code. This API is already pretty confusing and think that the inclusion of the Descriptor in this convenience class doesn't mean this class helps that very much. If you still have to know to query the Descriptor for types and codecs then what benefit does the rest of the Info class have over just using the AttributeSet itself.

I imagine in merging you want to merge Infos and so Descriptors need to be merged in the current setup so removing them from this struct would simplify that process completely no?
